Firestore/Source/Local/FSTPersistence.h

```diff
@@ -99,6 +100,32 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (void)commitGroup:(FSTWriteGroup *)group;
 
+@property (nonatomic, readonly, assign) const FSTTransactionRunner& run;
+
 @end
 
+@protocol FSTTransactional
+
+- (void)startTransaction;
+
+- (void)commitTransaction;
+
+@end
+
+struct FSTTransactionRunner {
+  template <typename F>
```

Comment:
I absolutely hate to do this, but this declaration does not work if block returns void. This is because you're assigning the value to `result` but you can't assign void to anything.

The way to fix this is to supply two implementations, one for void-returning blocks and one for not. You can control which one via some serious black magic like this:

```c++
  auto operator() (F block)
      -> typename std::enable_if<std::is_void<decltype(block())>::value, void>::type {
    block();
  }

  template <typename F>
  auto operator() (F block)
      -> typename std::enable_if<!std::is_void<decltype(block())>::value, decltype(block())>::type {
    auto result = block();
    return result;
  }
```

The principle that's at work here is that the C++ compiler obligated to ignore parameterized overloads whose types are nonsensical, but once the type is established to be sane actually error if the definition causes failures.

Briefly what's going on here is `std::enable_if` will expand the type in the second parameter if the expression in the first parameter is true. If the first parameter is false then it expands to an invalid type thus excluding the overload.

`std::is_void<T>::value` is true if the type is void.

Taken together the first overload is enabled if the return type of the block is `void`. The second is enabled if the return type is not `void`.

Please add comments :-).

Reply:
Hmm, actually, can this technique be extended to distinguish pointer and value types? That might be a better solution to the nullability issue, rather than just blindly suppressing the warning.

Reply:
While we can distinguish between pointer and non-pointer types (there's even built-in support for it via [`std::is_pointer`](http://en.cppreference.com/w/cpp/types/is_pointer), I don't think we can't distinguish between nullable and non-nullable pointers. This code passes:

```c++
static_assert(std::is_same<void* _Nullable, void*>::value, "nullable");
```

If you only want to support nullable or non-nullable types then you could add the following:

```c++
  template <typename F>
  auto operator() (F block)
      -> typename std::enable_if<
          std::is_pointer<decltype(block())>::value,
          decltype(block()) _Nullable>::type {
    auto&& result = block();
    return result;
  }

  template <typename F>
  auto operator() (F block)
      -> typename std::enable_if<
          !std::is_void<decltype(block())>::value &&
          !std::is_pointer<decltype(block())>::value,
          decltype(block())>::type {
    auto&& result = block();
    return result;
  }
```
Realistically though, I'm not sure this is any better than the version you have already.

Reply:
If I'm understanding correctly, we'd need the cross-product of void/non-void return type and pointer/non-pointer argument type. I agree that's overkill. I'll leave the two to handle the `void` case.